Allow representations with same name, fixes #8509

MaxNoe opened the pull request:
Fallback to `__qualname__` to avoid naming conflicts
in case there are multiple representations with
the same name.

fix #8509

mhvk commented:
See https://github.com/astropy/astropy/issues/8509#issuecomment-481257138 and following for some initial discussion.

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -436,8 +439,11 @@ def __init__(cls, name, bases, dct):
         if repr_name in REPRESENTATION_CLASSES:
-            raise ValueError("Representation class {0} already defined"
-                             .format(repr_name))
+            msg = 'Representation {} already defined, using name "{}"'.format(
```
```suggestion
            msg = 'Representation {} already defined, storing new representation with name "{}".  Note that this may cause surprising behavior if the representations are actually different!'.format(
```

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -436,8 +439,11 @@ def __init__(cls, name, bases, dct):
+            msg = 'Representation {} already defined, using name "{}"'.format(
+                repr_name, cls.__qualname__
+            )
+            warnings.warn(msg, DuplicateRepresentationWarning)
```
I think you still need to allow for the possibility that even the `__qualname__`'s are the same. So that would be something like:

```suggestion
            if cls.__qualname__ in REPRESENTATION_CLASSES:
                raise ValueError("Representation class {0} already defined".format(cls.__qualname__)
            warnings.warn(msg, DuplicateRepresentationWarning)
```

This is more of an edge-case than the previous one, but it's still possible when weird import things are happening and we may as well catch it just in case.

mhvk commented:
As noted in #8509, I'm not sure about this approach: I don't see the benefit of providing the representation with a different name - that means any code of the package that defined it will now use the wrong representation internally if it ever refers to it by name. Instead, I think one should either raise (as done currently) or warn and *remove* the one that is already in place (and thus remove any temptation to guess).

mhvk commented:
I removed the milestone to ensure this doesn't get merged accidentally.

eteq commented:
> I'm not sure about this approach

Just to be clear, you mean you don't want there to be *any* solution to #8509 ?  I think the fact that the answer to a legit user issue was "hack it by using some private variables" is a sign that we need *some* solution to #8509.  I think the best solution is indeed to try to get the upstream packages to not use overlapping identical names, but we can't really control that, so I think this solution is a way to work around it.

So when you say "remove the one that is already in place", do you mean make it so that string usage ends up not working for *either* of them?  I can see the argument for that (especially with your excellent Zen-of-Py-dropping :wink:), but won't that still break the upstream libraries in the way you're concerned about anyway? (or maybe you're saying that's *desirable*?)

mhvk commented:
I agree keeping the status quo is far from ideal... But better than an answer even possibly depending on import order of modules; that is just terrible!

An argument for instead choosing to drop both is based on the question @MaxNoe asked: why do we have the strings in the first place when the classes suffice? The representation classes are unique so if the actual modules only use those, the modules will continue to work fine even if the names are removed for both and there is no risk of interference. So, the removal proposal means that for two well-written modules, you get a warning but there is no risk of import-order-dependent answers, and for not so well-written modules, you get a warning that explains the possible error message about not finding a representation by name later on (which, as you suggest, I think is desirable...).

p.s. Note that we have the same problem for `Time` formats... And I realize I have used similar schemes in other packages...

MaxNoe commented:
I also thought about removing the short version for the first of the duplicated names and adding the qualname.

My concern was that then in the case of a third version, the duplicate wouldn't be recognized. But maybe this is so unlikely, that we can just ignore this?

What are you or opinions of fixing this by adding the planar representation creating the problems here to astropy? 

The different frames using this in the cherenkov astronomy use case are all detector coordinates, but there are multiple different conventions how the coordinate system is oriented towards the offset frame we call telescope frame, which is an altaz offset frame with the telescope pointing position as origin.

mhvk commented:
@MaxNoe - I think we'd need to have a `_duplicate` set which we also check against...

As for including this in astropy, I feel we clearly should, since it is done a lot. Indeed, effectively it is part of `wcs` already, but we somehow need to integrate it. I'd hope this would be do-able for 4.0.

eteq commented:
I also have no objection to including that representation in astropy, *but* I think we still want to address the underlying issue of this PR.  That is, if it happened here it might happen again for some other pair of packages down the road in some other sub-field...

eteq commented:
On this PR itself, I can see both sides now: the fact that upstream code gets broken when there's *any* namespace collision still creates a bit of an import-order-dependency problem (or "import at all" problem, really).  But I can see how it might be cleanest to refuse the temptation to guess.  @adrn do you have any thoughts here?

mhvk commented:
Re-opening since we should really get something like this in.

MaxNoe commented:
I rebased against the current master and implemented a ValueError if classes have the same qualname.

The behaviour is now:
* if duplicate name but different qualname: remove the short name and add both qualnames
* If duplicate name and same qualname: raise

This does not cover a case, where a *third* representation with the same name is defined. This would be again added normally. Do we want to deal with that somehow? E.g. tracking a list of duplicate names?

MaxNoe commented:
Sorry, @mhvk, rereading the older comments, you already proposed a duplicates set.

bsipocz commented:
@mhvk - is there a chance that this could still wrapped up? If not, please remilestone.

mhvk commented:
Darn, I should have looked at this earlier - @MaxNoe had already done everything I asked!

bsipocz commented:
Oh, if it's a bugfix then we're still OK, those will be backported even after the branching is done. Also, if it's only the changelog that needs fixing, I'm happy to do it prior/after branching/RC release, etc.

bsipocz commented:
Thank you @MaxNoe!
